File: app/src/main/java/solonsky/signal/twitter/fragments/DirectFragment.java
package solonsky.signal.twitter.fragments;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.anupcowkur.reservoir.Reservoir;
import com.anupcowkur.reservoir.ReservoirGetCallback;
import com.google.common.reflect.TypeToken;

import org.joda.time.LocalDateTime;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import solonsky.signal.twitter.R;
import solonsky.signal.twitter.activities.ChatActivity;
import solonsky.signal.twitter.activities.MVPProfileActivity;
import solonsky.signal.twitter.adapters.DirectAdapter;
import solonsky.signal.twitter.api.DirectApi;
import solonsky.signal.twitter.data.DirectData;
import solonsky.signal.twitter.data.LoggedData;
import solonsky.signal.twitter.databinding.FragmentDirectBinding;
import solonsky.signal.twitter.helpers.App;
import solonsky.signal.twitter.helpers.AppData;
import solonsky.signal.twitter.helpers.Cache;
import solonsky.signal.twitter.helpers.Flags;
import solonsky.signal.twitter.interfaces.ActivityListener;
import solonsky.signal.twitter.interfaces.UpdateHandler;
import solonsky.signal.twitter.helpers.Utilities;
import solonsky.signal.twitter.models.DirectModel;
import solonsky.signal.twitter.models.User;
import solonsky.signal.twitter.viewmodels.DirectViewModel;
import twitter4j.AsyncTwitter;
import twitter4j.DirectMessage;
import twitter4j.Paging;
import twitter4j.ResponseList;
import twitter4j.TwitterAdapter;
import twitter4j.TwitterException;
import twitter4j.TwitterMethod;

/**
 * Created by neura on 23.05.17.
 */

public class DirectFragment extends Fragment {
    private final String TAG = DirectFragment.class.getSimpleName();
    private int newCount = 0;
    private DirectAdapter mAdapter;
    private DirectViewModel viewModel;
    private ActivityListener mCallback;

    private RecyclerView.OnScrollListener directScrollListener = new RecyclerView.OnScrollListener() {
        @Override
        public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
            super.onScrolled(recyclerView, dx, dy);
            if (mCallback != null) {
                mCallback.updateBars(dy);
            }
        }
    };

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        FragmentDirectBinding binding = DataBindingUtil.inflate(inflater, R.layout.fragment_direct, container, false);

        mAdapter = new DirectAdapter(DirectData.getInstance().getmMessagesList(), getContext(), (AppCompatActivity) getActivity(),
                new DirectAdapter.DirectClickHandler() {
                    @Override
                    public void onItemClick(View v, final DirectModel model) {
                        AppData.DM_POSITION = DirectData.getInstance().getmMessagesList().indexOf(model);
                        AppData.DM_SELECTED_USER = model.getUsername();
                        AppData.DM_OTHER_ID = model.getOtherId();
                        Flags.DM_IS_NEW = model.getMessageCount() > 0;

                        DirectApi.getInstance().clear();
                        DirectApi.getInstance().setUserId(model.getOtherId());
                        getActivity().startActivity(new Intent(getContext(), ChatActivity.class));
                        if (model.isHighlighted()) {
                            getActivity().overridePendingTransition(R.anim.slide_in_up, R.anim.slide_out_no_animation);
                        } else {
                            getActivity().overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left);
                        }

                        new Handler().postDelayed(new Runnable() {
                            @Override
                            public void run() {
                                model.setHighlighted(false);

                                model.setMessageCount(0);
                                boolean hasNew = false;
                                for (DirectModel directModel : DirectData.getInstance().getmMessagesList()) {
                                    if (directModel.getMessageCount() > 0) {
                                        hasNew = true;
                                        break;
                                    }
                                }

                                if (!hasNew) {
                                    LoggedData.getInstance().setNewMessage(false);
                                    LoggedData.getInstance().getUpdateHandler().onUpdate();
                                }

                                saveCache();
                            }
                        }, 300);
                    }

                    @Override
                    public void onAvatarClick(View v, DirectModel directModel) {
                        Intent profileIntent = new Intent(getContext(), MVPProfileActivity.class);
                        profileIntent.putExtra(Flags.PROFILE_ID, directModel.getOtherId());
                        getActivity().startActivity(profileIntent);
                        getActivity().overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left);
                    }
                });
        mAdapter.setHasStableIds(true);

        if (DirectData.getInstance().getmMessagesList().size() == 0) {
            viewModel = new DirectViewModel(mAdapter, getContext(), AppData.UI_STATE_LOADING);
        } else {
            viewModel = new DirectViewModel(mAdapter, getContext(), AppData.UI_STATE_VISIBLE);
        }
        binding.setModel(viewModel);

        if (mCallback != null) {
            mCallback.updateCounter(DirectData.getInstance().getEntryCount());
            mCallback.updateSettings(R.string.title_direct, false, false);
            mCallback.updateToolbarState(AppData.TOOLBAR_LOGGED_DM, App.getInstance().isNightEnabled() ?
                    R.color.dark_status_bar_timeline_color : R.color.light_status_bar_timeline_color);
        }

        boolean isNight = App.getInstance().isNightEnabled();

//        for (int i = 0; i < mMessagesList.size(); i++) {
//            if (i == mMessagesList.size() - 1) {
//                mMessagesList.get(i).setDivideState(AppData.DIVIDER_LONG);
//            } else {
//                mMessagesList.get(i).setDivideState(
//                        mMessagesList.get(i).isHighlighted() && mMessagesList.get(i + 1).isHighlighted()
//                                || !mMessagesList.get(i).isHighlighted() && !mMessagesList.get(i + 1).isHighlighted()
//                                ? AppData.DIVIDER_SHORT : isNight ? AppData.DIVIDER_NONE : AppData.DIVIDER_LONG);
//            }
//        }

        binding.recyclerDirect.setOnScrollListener(directScrollListener);
        DirectData.getInstance().setUpdateHandler(new UpdateHandler() {
            @Override
            public void onUpdate() {
                mAdapter.notifyDataSetChanged();
                saveCache();
            }
        });

        if (DirectData.getInstance().getmMessagesList().size() == 0) {
            loadCache();
        }
        return binding.getRoot();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (mAdapter != null) {
            if (isAdded() && Flags.needsToRedrawDirect) {
                Flags.needsToRedrawDirect = false;
                mAdapter.notifyDataSetChanged();
            }

            if (Flags.isUpdated) {
                mAdapter.notifyItemRangeChanged(0, mAdapter.getItemCount());
            }
            if (Flags.DELETE_THREAD) {
                Flags.DELETE_THREAD = false;
                mAdapter.notifyItemRemoved(AppData.DM_POSITION);
            }
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        mCallback = (ActivityListener) context;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mCallback = null;
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        if (!hidden) {
            if (mCallback != null) {
                mCallback.updateCounter(DirectData.getInstance().getEntryCount());
                mCallback.updateSettings(R.string.title_direct, true, true);
                mCallback.updateToolbarState(AppData.TOOLBAR_LOGGED_DM, App.getInstance().isNightEnabled() ?
                        R.color.dark_status_bar_timeline_color : R.color.light_status_bar_timeline_color);
            }
        }
    }

    /**
     * Load cache from disk
     *
     * @serialData tweets array
     */
    private void loadCache() {
        Type resultType = new TypeToken<List<DirectModel>>() {
        }.getType();
        Reservoir.getAsync(Cache.DirectsList + String.valueOf(AppData.ME.getId()), resultType,
                new ReservoirGetCallback<List<DirectModel>>() {
                    @Override
                    public void onFailure(Exception e) {
                        loadApi(new Paging(1, 200));
                    }

                    @Override
                    public void onSuccess(List<DirectModel> directModels) {
                        for (DirectModel directModel : directModels) {
                            DirectData.getInstance().getmMessagesList().add(directModel);
                        }

                        mAdapter.notifyDataSetChanged();
                        viewModel.setState(DirectData.getInstance().getmMessagesList().size() == 0 ?
                                AppData.UI_STATE_NO_ITEMS : AppData.UI_STATE_VISIBLE);
                        loadApi(new Paging(1, 200));
                    }
                });
    }

    /**
     * Save cache
     *
     * @serialData bytecode to disk
     */
    private void saveCache() {
        try {
            Reservoir.put(Cache.DirectsList + String.valueOf(AppData.ME.getId()), DirectData.getInstance().getmMessagesList());
        } catch (IOException e) {
            Log.e(TAG, "Error caching directs - " + e.getLocalizedMessage());
        }
    }

    /**
     * Loading direct messages from twitter backend
     */
    public void loadApi(final Paging paging) {
        final Handler handler = new Handler();
        final ArrayList<twitter4j.DirectMessage> loadedDirects = new ArrayList<>();
        final AsyncTwitter asyncTwitter = Utilities.getAsyncTwitter();
        asyncTwitter.addListener(new TwitterAdapter() {
            @Override
            public void gotDirectMessages(ResponseList<twitter4j.DirectMessage> messages) {
                super.gotDirectMessages(messages);
                loadedDirects.addAll(messages);
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        asyncTwitter.getSentDirectMessages(paging);
                    }
                });
            }

            @Override
            public void gotSentDirectMessages(ResponseList<twitter4j.DirectMessage> messages) {
                super.gotSentDirectMessages(messages);
                loadedDirects.addAll(messages);

                Collections.sort(loadedDirects, new Comparator<twitter4j.DirectMessage>() {
                    @Override
                    public int compare(twitter4j.DirectMessage o1, twitter4j.DirectMessage o2) {
                        return o2.getCreatedAt().compareTo(o1.getCreatedAt());
                    }
                });

                List<twitter4j.DirectMessage> newDirects = new ArrayList<>();
                List<DirectMessage> newUserDirects = new ArrayList<>();

                if (DirectData.getInstance().getmMessagesList().size() == 0) {
                    for (twitter4j.DirectMessage directMessage : loadedDirects) {
                        long id = directMessage.getSenderId() == AppData.ME.getId() ? directMessage.getRecipientId() : directMessage.getSenderId();
                        solonsky.signal.twitter.models.User user = User.getFromUserInstance(directMessage.getSenderId() == AppData.ME.getId() ?
                                directMessage.getRecipient() : directMessage.getSender());

                        DirectModel directModel = new DirectModel(directMessage.getId(), id, user.getOriginalProfileImageURL(),
                                user.getName(), directMessage.getText(), new LocalDateTime(directMessage.getCreatedAt()),
                                false);
                        if (!DirectData.getInstance().getmMessagesList().contains(directModel)) {
                            DirectData.getInstance().getmMessagesList().add(directModel);
                        }
                    }
                } else {
                    for (twitter4j.DirectMessage directMessage : loadedDirects) {
                        boolean hasId = false;
                        int position = 0;
                        for (DirectModel directModel : DirectData.getInstance().getmMessagesList()) {
                            if (directMessage.getSenderId() == directModel.getOtherId() ||
                                    directMessage.getRecipientId() == directModel.getOtherId()) {
                                hasId = true;
                                position = DirectData.getInstance().getmMessagesList().indexOf(directModel);
                                break;
                            }
                        }

                        if (hasId) {
                            if (directMessage.getId() > DirectData.getInstance().getmMessagesList().get(position).getId()) {
                                newDirects.add(directMessage);
                            }
                        } else {
                            newUserDirects.add(directMessage);
                        }
                    }

                    for (twitter4j.DirectMessage directMessage : newDirects) {
                        long id = AppData.ME.getId() == directMessage.getSenderId() ?
                                directMessage.getRecipientId() : directMessage.getSenderId();

                        solonsky.signal.twitter.models.User user = AppData.ME.getId() == directMessage.getSenderId() ?
                                User.getFromUserInstance(directMessage.getRecipient()) :
                                User.getFromUserInstance(directMessage.getSender());

                        int entryCount = 0;
                        for (DirectModel directModel : DirectData.getInstance().getmMessagesList()) {
                            if (directModel.getOtherId() == directMessage.getSenderId()
                                    || directModel.getOtherId() == directMessage.getRecipientId()) {
                                entryCount = directModel.getMessageCount();
                                DirectData.getInstance().getmMessagesList().remove(
                                        DirectData.getInstance().getmMessagesList().indexOf(directModel));
                                break;
                            }
                        }

                        DirectModel directModel = new DirectModel(
                                directMessage.getId(), id, user.getOriginalProfileImageURL(),
                                user.getName(), directMessage.getText(), new LocalDateTime(directMessage.getCreatedAt()), false);
                        if (directMessage.getSenderId() != AppData.ME.getId())
                            directModel.setMessageCount(entryCount + 1);
                        DirectData.getInstance().getmMessagesList().add(0, directModel);
                    }

                    for (DirectMessage directMessage : newUserDirects) {
                        long id = AppData.ME.getId() == directMessage.getSenderId() ?
                                directMessage.getRecipientId() : directMessage.getSenderId();
                        boolean hasId = false;

                        for (DirectModel directModel : DirectData.getInstance().getmMessagesList()) {
                            if (directModel.getOtherId() == id) {
                                hasId = true;
                                break;
                            }
                        }

                        if (!hasId) {
                            twitter4j.User parsedUser = AppData.ME.getId() == directMessage.getSenderId() ?
                                    directMessage.getRecipient() : directMessage.getSender();
                            User user = User.getFromUserInstance(parsedUser);
                            DirectModel directModel = new DirectModel(
                                    directMessage.getId(), id, user.getOriginalProfileImageURL(),
                                    user.getName(), directMessage.getText(), new LocalDateTime(directMessage.getCreatedAt()), false);
                            if (!DirectData.getInstance().getmMessagesList().contains(directModel)) {
                                DirectData.getInstance().getmMessagesList().add(directModel);
                            }
                        }
                    }
                }

                boolean hasNew = false;
                for (DirectModel directModel : DirectData.getInstance().getmMessagesList()) {
                    if (directModel.getMessageCount() > 0) {
                        hasNew = true;
                        break;
                    }
                }

                LoggedData.getInstance().setNewMessage(hasNew);

                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        mAdapter.notifyDataSetChanged();
                        viewModel.setState(DirectData.getInstance().getmMessagesList().size() == 0 ?
                                AppData.UI_STATE_NO_ITEMS : AppData.UI_STATE_VISIBLE);
                        LoggedData.getInstance().getUpdateHandler().onUpdate();
                    }
                });

//                saveCache();
            }

            @Override
            public void onException(TwitterException te, TwitterMethod method) {
                super.onException(te, method);
                Log.e(TAG, "Error while loading data - " + te.getLocalizedMessage());
//                Toast.makeText(getContext(), "Error while loading data - " + te.getLocalizedMessage(), Toast.LENGTH_SHORT).show();
            }
        });

        asyncTwitter.getDirectMessages(paging);
    }
}

File: app/src/main/java/solonsky/signal/twitter/fragments/DirectFragment.kt
package solonsky.signal.twitter.fragments

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView

import com.anupcowkur.reservoir.Reservoir
import com.anupcowkur.reservoir.ReservoirGetCallback
import com.google.common.reflect.TypeToken

import org.joda.time.LocalDateTime

import java.io.IOException

import solonsky.signal.twitter.R
import solonsky.signal.twitter.activities.ChatActivity
import solonsky.signal.twitter.activities.MVPProfileActivity
import solonsky.signal.twitter.adapters.DirectAdapter
import solonsky.signal.twitter.api.DirectApi
import solonsky.signal.twitter.data.DirectData
import solonsky.signal.twitter.data.LoggedData
import solonsky.signal.twitter.databinding.FragmentDirectBinding
import solonsky.signal.twitter.helpers.*
import solonsky.signal.twitter.interfaces.ActivityListener
import solonsky.signal.twitter.interfaces.UpdateHandler
import solonsky.signal.twitter.models.DirectModel
import solonsky.signal.twitter.models.User
import solonsky.signal.twitter.viewmodels.DirectViewModel
import twitter4j.*
import java.util.*

/**
 * Created by neura on 23.05.17.
 */

class DirectFragment : Fragment() {
    private val TAG = DirectFragment::class.java.simpleName
    private val newCount = 0
    private lateinit var mAdapter: DirectAdapter
    private lateinit var binding: FragmentDirectBinding
    private var viewModel: DirectViewModel? = null
    private var mCallback: ActivityListener? = null

    private val directScrollListener = object : RecyclerView.OnScrollListener() {
        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
            super.onScrolled(recyclerView, dx, dy)
            mCallback?.let { it.updateBars(dy) }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = DataBindingUtil.inflate(inflater!!, R.layout.fragment_direct, container, false)

        mAdapter = DirectAdapter(DirectData.getInstance().getmMessagesList(), context, activity as AppCompatActivity,
                object : DirectAdapter.DirectClickHandler {
                    override fun onItemClick(v: View, model: DirectModel) {
                        AppData.DM_POSITION = DirectData.getInstance().getmMessagesList().indexOf(model)
                        AppData.DM_SELECTED_USER = model.username
                        AppData.DM_OTHER_ID = model.otherId
                        Flags.DM_IS_NEW = model.messageCount > 0

                        DirectApi.getInstance().clear()
                        DirectApi.getInstance().userId = model.otherId
                        (activity as AppCompatActivity).startActivity(Intent(context, ChatActivity::class.java))
                        if (model.isHighlighted) {
                            (activity as AppCompatActivity).overridePendingTransition(R.anim.slide_in_up, R.anim.slide_out_no_animation)
                        } else {
                            (activity as AppCompatActivity).overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left)
                        }

                        Handler().postDelayed({
                            model.isHighlighted = false
                            model.messageCount = 0
                            val hasNew = DirectData.getInstance().getmMessagesList().any { it.messageCount > 0 }

                            if (!hasNew) {
                                LoggedData.getInstance().isNewMessage = false
                                LoggedData.getInstance().updateHandler.onUpdate()
                            }

                            saveCache()
                        }, 300)
                    }

                    override fun onAvatarClick(v: View, directModel: DirectModel) {
                        val profileIntent = Intent(context, MVPProfileActivity::class.java)
                        profileIntent.putExtra(Flags.PROFILE_ID, directModel.otherId)
                        (activity as AppCompatActivity).startActivity(profileIntent)
                        (activity as AppCompatActivity).overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left)
                    }
                })
        mAdapter.setHasStableIds(true)

        viewModel = if (DirectData.getInstance().getmMessagesList().size == 0) {
            DirectViewModel(mAdapter, context, AppData.UI_STATE_LOADING)
        } else {
            DirectViewModel(mAdapter, context, AppData.UI_STATE_VISIBLE)
        }

        binding.model = viewModel

        mCallback?.let {
            it.updateCounter(DirectData.getInstance().entryCount)
            it.updateSettings(R.string.title_direct, false, false)
            it.updateToolbarState(AppData.TOOLBAR_LOGGED_DM, if (App.getInstance().isNightEnabled)
                R.color.dark_status_bar_timeline_color
            else
                R.color.light_status_bar_timeline_color)
        }

        binding.recyclerDirect.setOnScrollListener(directScrollListener)
        binding.srlDirectMain.setProgressViewOffset(false,
                Utilities.convertDpToPixel(80f, context).toInt(),
                Utilities.convertDpToPixel(96f, context).toInt())
        binding.srlDirectMain.setOnRefreshListener {
            refreshData()
        }

        DirectData.getInstance().updateHandler = UpdateHandler {
            mAdapter.notifyDataSetChanged()
            saveCache()
        }

        if (DirectData.getInstance().getmMessagesList().size == 0) {
            loadCache()
        }
        return binding.root
    }

    override fun onResume() {
        super.onResume()
        try {
            if (isAdded && Flags.needsToRedrawDirect) {
                Flags.needsToRedrawDirect = false
                mAdapter.notifyDataSetChanged()
            }

            if (Flags.isUpdated) {
                mAdapter.notifyItemRangeChanged(0, mAdapter.itemCount)
            }
            if (Flags.DELETE_THREAD) {
                Flags.DELETE_THREAD = false
                mAdapter.notifyItemRemoved(AppData.DM_POSITION)
            }
        } catch (e: NullPointerException) {
            e.printStackTrace()
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        mCallback = context as ActivityListener?
    }

    override fun onDetach() {
        super.onDetach()
        mCallback = null
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        if (!hidden) {
            mCallback?.let {
                it.updateCounter(DirectData.getInstance().entryCount)
                it.updateSettings(R.string.title_direct, true, true)
                it.updateToolbarState(AppData.TOOLBAR_LOGGED_DM, if (App.getInstance().isNightEnabled)
                    R.color.dark_status_bar_timeline_color
                else
                    R.color.light_status_bar_timeline_color)
            }
        }
    }

    /**
     * Load cache from disk
     *
     * @serialData tweets array
     */
    private fun loadCache() {
        val resultType = object : TypeToken<List<DirectModel>>() {}.type
        Reservoir.getAsync(Cache.DirectsList + AppData.ME.id.toString(), resultType,
                object : ReservoirGetCallback<List<DirectModel>> {
                    override fun onFailure(e: Exception) {
                        loadApi(Paging(1, 200))
                    }

                    override fun onSuccess(directModels: List<DirectModel>) {
                        for (directModel in directModels) {
                            DirectData.getInstance().getmMessagesList().add(directModel)
                        }

                        mAdapter.notifyDataSetChanged()
                        (if (DirectData.getInstance().getmMessagesList().size == 0)
                            AppData.UI_STATE_NO_ITEMS
                        else
                            AppData.UI_STATE_VISIBLE).also { viewModel!!.state = it }
                        loadApi(Paging(1, 200))
                    }
                })
    }

    /**
     * Save cache
     *
     * @serialData bytecode to disk
     */
    private fun saveCache() {
        try {
            Reservoir.put(Cache.DirectsList + AppData.ME.id.toString(), DirectData.getInstance().getmMessagesList())
        } catch (e: IOException) {
            Log.e(TAG, "Error caching directs - " + e.localizedMessage)
        }

    }

    private fun refreshData() = if (DirectData.getInstance().getmMessagesList().size > 0) {
        val handler = Handler()
        val asyncTwitter = Utilities.getAsyncTwitter()
        val paging = Paging()
        paging.sinceId = DirectData.getInstance().getmMessagesList()[0].id

        asyncTwitter.addListener(object: TwitterAdapter() {
            val directDictionary: HashMap<Long, DirectMessage> = HashMap()

            override fun onException(te: TwitterException?, method: TwitterMethod?) {
                super.onException(te, method)
                handler.post {
                    Toast.makeText(context, getString(R.string.error_loading_direct), Toast.LENGTH_SHORT).show()
                    binding.srlDirectMain.isRefreshing = false
                }
            }

            override fun gotSentDirectMessages(messages: ResponseList<DirectMessage>?) {
                super.gotSentDirectMessages(messages)
                messages?.reverse()
                messages?.forEach {
                    val id = if (it.senderId == AppData.ME.id)
                        it.recipientId
                    else
                        it.senderId

                    if (directDictionary.containsKey(id)) {
                        directDictionary.remove(id)
                    }

                    directDictionary.put(id, it)
                }

                Log.e(TAG, "size ${directDictionary.size}")

                DirectData.getInstance().getmMessagesList().forEach {
                    if (directDictionary.containsKey(it.otherId)) {
                        Log.e(TAG, "contains")
                        directDictionary[it.otherId]?.let { direct ->
                            Log.e(TAG, "message ${direct.text}")
                            it.lastMessage = direct.text
                            it.isHighlighted = true
                        }
                        directDictionary.remove(it.otherId)
                    }
                }

                directDictionary.keys.forEach {
                    DirectData.getInstance().getmMessagesList()
                        .add(DirectModel.getInstance(directDictionary[it], AppData.ME.id))
                }

                handler.post {
                    binding.srlDirectMain.isRefreshing = false
                    mAdapter.notifyDataSetChanged()
                }
            }

            override fun gotDirectMessages(messages: ResponseList<DirectMessage>?) {
                super.gotDirectMessages(messages)
                messages?.reverse()
                messages?.forEach {
                    val id = if (it.senderId == AppData.ME.id)
                        it.recipientId
                    else
                        it.senderId

                    if (directDictionary.containsKey(id)) {
                        directDictionary.remove(id)
                    }

                    directDictionary[id] = it
                }

                handler.post {
                    asyncTwitter.getSentDirectMessages(paging)
                }
            }
        })

        asyncTwitter.getDirectMessages(paging)
    } else {
        binding.srlDirectMain.isRefreshing = false
    }

    /**
     * Loading direct messages from twitter backend
     */
    fun loadApi(paging: Paging) {
        val handler = Handler()
        val loadedDirects = ArrayList<twitter4j.DirectMessage>()
        val asyncTwitter = Utilities.getAsyncTwitter()
        asyncTwitter.addListener(object : TwitterAdapter() {
            override fun gotDirectMessages(messages: ResponseList<twitter4j.DirectMessage>) {
                super.gotDirectMessages(messages)
                loadedDirects.addAll(messages)
                handler.post { asyncTwitter.getSentDirectMessages(paging) }
            }

            override fun gotSentDirectMessages(messages: ResponseList<twitter4j.DirectMessage>) {
                super.gotSentDirectMessages(messages)
                loadedDirects.addAll(messages)

                loadedDirects.sortWith(Comparator { o1, o2 -> o2.createdAt.compareTo(o1.createdAt) })

                val newDirects = ArrayList<twitter4j.DirectMessage>()
                val newUserDirects = ArrayList<DirectMessage>()

                if (DirectData.getInstance().getmMessagesList().size == 0) {
                    for (directMessage in loadedDirects) {
                        val id = if (directMessage.senderId == AppData.ME.id) directMessage.recipientId else directMessage.senderId
                        val user = User.getFromUserInstance(if (directMessage.senderId == AppData.ME.id)
                            directMessage.recipient
                        else
                            directMessage.sender)

                        val text = directMessage.text
                        directMessage.urlEntities.forEach {
                            text.replace(it.url, it.displayURL)
                        }

                        val directModel = DirectModel(directMessage.id, id, user.originalProfileImageURL,
                                user.name, text, LocalDateTime(directMessage.createdAt),
                                false)
                        if (!DirectData.getInstance().getmMessagesList().contains(directModel)) {
                            DirectData.getInstance().getmMessagesList().add(directModel)
                        }
                    }
                } else {
                    for (directMessage in loadedDirects) {
                        var hasId = false
                        var position = 0
                        for (directModel in DirectData.getInstance().getmMessagesList()) {
                            if (directMessage.senderId == directModel.otherId || directMessage.recipientId == directModel.otherId) {
                                hasId = true
                                position = DirectData.getInstance().getmMessagesList().indexOf(directModel)
                                break
                            }
                        }

                        if (hasId) {
                            if (directMessage.id > DirectData.getInstance().getmMessagesList()[position].id) {
                                newDirects.add(directMessage)
                            }
                        } else {
                            newUserDirects.add(directMessage)
                        }
                    }

                    for (directMessage in newDirects) {
                        val id = if (AppData.ME.id == directMessage.senderId)
                            directMessage.recipientId
                        else
                            directMessage.senderId

                        val user = if (AppData.ME.id == directMessage.senderId)
                            User.getFromUserInstance(directMessage.recipient)
                        else
                            User.getFromUserInstance(directMessage.sender)

                        var entryCount = 0
                        for (directModel in DirectData.getInstance().getmMessagesList()) {
                            if (directModel.otherId == directMessage.senderId || directModel.otherId == directMessage.recipientId) {
                                entryCount = directModel.messageCount
                                DirectData.getInstance().getmMessagesList().removeAt(
                                        DirectData.getInstance().getmMessagesList().indexOf(directModel))
                                break
                            }
                        }

                        val text = directMessage.text
                        directMessage.urlEntities.forEach {
                            Log.e(TAG, "before replace $text")
                            text.replace(it.url, it.displayURL)
                            Log.e(TAG, "after replace $text")
                        }

                        val directModel = DirectModel(
                                directMessage.id, id, user.originalProfileImageURL,
                                user.name, text, LocalDateTime(directMessage.createdAt), false)
                        if (directMessage.senderId != AppData.ME.id)
                            directModel.messageCount = entryCount + 1
                        DirectData.getInstance().getmMessagesList().add(0, directModel)
                    }

                    for (directMessage in newUserDirects) {
                        val id = if (AppData.ME.id == directMessage.senderId)
                            directMessage.recipientId
                        else
                            directMessage.senderId
                        val hasId = DirectData.getInstance().getmMessagesList().any { it.otherId == id }

                        if (!hasId) {
                            val parsedUser = if (AppData.ME.id == directMessage.senderId)
                                directMessage.recipient
                            else
                                directMessage.sender
                            val user = User.getFromUserInstance(parsedUser)

                            val text = directMessage.text
                            directMessage.urlEntities.forEach {
                                text.replace(it.url, it.displayURL)
                            }

                            val directModel = DirectModel(
                                    directMessage.id, id, user.originalProfileImageURL,
                                    user.name, text, LocalDateTime(directMessage.createdAt), false)
                            if (!DirectData.getInstance().getmMessagesList().contains(directModel)) {
                                DirectData.getInstance().getmMessagesList().add(directModel)
                            }
                        }
                    }
                }

                val hasNew = DirectData.getInstance().getmMessagesList().any { it.messageCount > 0 }
                LoggedData.getInstance().isNewMessage = hasNew

                handler.post {
                    mAdapter.notifyDataSetChanged()
                    viewModel!!.state = if (DirectData.getInstance().getmMessagesList().size == 0)
                        AppData.UI_STATE_NO_ITEMS
                    else
                        AppData.UI_STATE_VISIBLE
                    LoggedData.getInstance().updateHandler.onUpdate()
                }
            }

            override fun onException(te: TwitterException?, method: TwitterMethod?) {
                super.onException(te, method)
                te?.let {
                    Log.e(TAG, "Error while loading data - " + it.localizedMessage)
                }
            }
        })

        asyncTwitter.getDirectMessages(paging)
    }
}
